### PR TITLE
Implementation of $backupCursor and $backupCursorExtend

### PR DESCRIPTION
**Pull request by igorsol**

### CODE REVIEW

File: src/mongo/db/pipeline/document_source_backup_cursor.cpp [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    if (_docIt != _backupInformation.end()) {
+        Document doc = {{"filename"_sd, _docIt->first},
+                        {"fileSize"_sd, static_cast<long long>(_docIt->second.fileSize)}};
+        ++_docIt;
+
+        return doc;
+    }
+
+    return GetNextResult::makeEOF();
+}
```
Comment (akira-kurogane) [inline src/mongo/db/pipeline/document_source_backup_cursor.cpp:93]: The backup cursor usage example in the catalog/README.md mentions that at a getmore command needs to be sent as a 'heartbeat' at least once every 10 mins to keep the backup cursor alive.

https://github.com/mongodb/mongo/blob/master/src/mongo/db/catalog/README.md#copy-files-while-sending-heartbeats

10 mins is the default for cursorTimeoutMillis, maybe the agg pipeline $backupCursor stage's cursor is just inheriting that and we don't need to add anything. But let's check that we can send a getmore without having an error, can do backups where one file takes > 10 mins to copy.

File: src/mongo/db/storage/wiredtiger/wiredtiger_kv_engine.cpp [hunk 2]
```diff
@@ -1554,6 +1558,98 @@ StatusWith<std::vector<std::string>> WiredTigerKVEngine::extendBackupCursor(
...
+    _backupCursor = cursor;
+
+    return swBackupInfo;
+}
+
+Status EncryptionKeyDB::endNonBlockingBackup() {
+    _backupSession.reset();
+    _backupCursor = nullptr;
+    return Status::OK();
+}
```
[the remark this answers is not in the document]
Comment (igorsol) [reply src/mongo/db/storage/wiredtiger/wiredtiger_kv_engine.cpp:1609]: Good catch. There is WiredTigerKVEngine::disableIncrementalBackup() and $backupCursor aggregation stage accepts "disableIncrementalBackup" boolean parameter but it seems that parameter is just ignored. I will continue investigation.

File: src/mongo/db/pipeline/document_source_backup_cursor.cpp (hunk 1 above)
Comment (igorsol) [reply src/mongo/db/pipeline/document_source_backup_cursor.cpp:93]: I tested getMore command command. It works without errors. The interesting part is "batchSize" parameter which exists in "getMore" and also in aggregation stage itself. I will test if "batchSize" is respected and will put update on that here.

Comment (igorsol) [reply src/mongo/db/pipeline/document_source_backup_cursor.cpp:93]: "batchSize" works as expected.

File: src/mongo/db/storage/wiredtiger/wiredtiger_kv_engine.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (igorsol) [reply src/mongo/db/storage/wiredtiger/wiredtiger_kv_engine.cpp:1666]: Added commit b4ddaec to handle 'disableIncrementalBackup'

[the remark this answers is not in the document]
Comment (igorsol) [reply src/mongo/db/storage/wiredtiger/wiredtiger_kv_engine.cpp:1666]: To disable incremental backup client program should:
- open backup cursor with 'disableIncrementalBackup=true'
- kill that cursor
```
> use admin
switched to db admin
> db.runCommand({aggregate:1, cursor: {}, pipeline: [{$backupCursor: {disableIncrementalBackup: true}}]})
{
	"cursor" : {
		"firstBatch" : [
			{
				"metadata" : {
					"backupId" : UUID("b8d6f54f-6474-4056-a55c-ece5e9347a23"),
					"dbpath" : "/psmdb-802/delme"
				}
			}
		],
		"id" : NumberLong("8062535432179657690"),
		"ns" : "admin.$cmd.aggregate"
	},
	"ok" : 1
}
> db.runCommand({killCursors: "admin.$cmd.aggregate", cursors: [NumberLong("8062535432179657690")]})
{
	"cursorsKilled" : [
		NumberLong("8062535432179657690")
	],
	"cursorsNotFound" : [ ],
	"cursorsAlive" : [ ],
	"cursorsUnknown" : [ ],
	"ok" : 1
}
```

**Review by akira-kurogane — APPROVED** [review]
Thanks Igor

**Pull request merged**